include/crm/common/action_relation_internal.h

```diff
@@ -0,0 +1,130 @@
+     * required (and becomes unmigratable if 'then' is)
+     */
+    pcmk__ar_promoted_then_implies_first    = (1U << 6),
+
+    /*!
+     * 'first' is runnable only if 'then' is both runnable and migratable,
+     * and 'first' becomes required if 'then' is
+     */
+    pcmk__ar_unmigratable_then_blocks       = (1U << 7),
+
+    //! 'then' is runnable (and migratable) only if 'first' is runnable
+    pcmk__ar_unrunnable_first_blocks        = (1U << 8),
+
+    //! If 'first' is unrunnable, 'then' becomes a real, unmigratable action
+    pcmk__ar_first_else_then                = (1U << 9),
+
+    //! If 'first' is required, 'then' action for instance on same node is
+    pcmk__ar_first_implies_same_node_then   = (1U << 10),
+
+    /*!
+     * Disable ordering if 'first' is unrunnable but for an already-active
+     * resource, otherwise 'then' is runnable only if 'first' is runnable.
+     *
+     * This is used to order a bundle replica's start of its container before a
+     * probe of its remote connection resource, in case the connection uses the
+     * REMOTE_CONTAINER_HACK to replace the connection address with where the
+     * container is running.
+     */
+    pcmk__ar_nested_remote_probe            = (1U << 11),
+
+    /*!
+     * If 'first' is for a blocked resource, make 'then' unrunnable.
+     *
+     * If 'then' is required, make 'first' required, make 'first' unmigratable
+     * if 'then' is unmigratable, and make 'then' unrunnable if 'first' is
+     * unrunnable.
+     *
+     * If 'then' is unrunnable and for the same resource as 'first', make
+     * 'first' required if it is runnable, and make 'first' unmigratable if
+     * 'then' is unmigratable.
+     *
+     * This is used for "stop then start primitive" (restarts) and
+     * "stop group member then stop previous member".
+     */
+    pcmk__ar_intermediate_stop              = (1U << 12),
+
+    /*!
+     * The actions must be serialized if in the same transition but can be in
+     * either order. (In practice, we always arrange them as 'first' then
+     * 'then', so they end up being essentially the same as optional orderings.)
+     *
+     * @TODO Handle more intelligently -- for example, we could schedule the
+     * action with the fewest inputs first, so we're more likely to execute at
+     * least one if there is a failure during the transition. Or, we could
+     * prefer certain action types over others, or base it on resource priority.
+     */
+    pcmk__ar_serialize                      = (1U << 14),
+
+    //! Relation applies only if actions are on same node
+    pcmk__ar_if_on_same_node                = (1U << 15),
+
+    //! If 'then' is required, 'first' must be added to the transition graph
+    pcmk__ar_then_implies_first_graphed     = (1U << 16),
+
+    //! If 'first' is required and runnable, 'then' must be in graph
+    pcmk__ar_first_implies_then_graphed     = (1U << 17),
+
+    //! User-configured asymmetric ordering
```

Comment:
This one could stay "ordering" or could become "relation" for future uses

Reply:
Yeah I think I'll leave it for the future. It really needs a comprehensive plan for how flags *should* work. I'm thinking asymmetric should be the default (in code) and a flag should be set for symmetric, since all internally created orderings are asymmetric.